python/imageutils.py

```diff
@@ -295,125 +185,55 @@ def reconstruct_block_slice(self, reconstructed, legend, mem):
+        total_write = 0
+
         with open(legend, "r") as legend:
 
             remaining_mem = mem
-            data = None
-            start_x = 0
-            start_z = 0
-            start_y = 0
+            data_dict = {}
 
-            prev_block_start = (0,0,0) 
-            for block_name in legend:
-                
-                block_name = block_name.strip()
-                print "Reading block {0}".format(block_name)
-                block_pos = split_ext(block_name)[0].split('_')
-
-                #we will start our (partial) slice at these block coordinates
-                if remaining_mem == mem:
-                    start_x = int(block_pos[-1])
-                    start_z = int(block_pos[-2])
-                    start_y = int(block_pos[-3])
-                    
-                block = nib.load(block_name)
-                block_shape = block.header.get_data_shape()
-                bytes_per_voxel_block = block.header['bitpix']/8
-                block_bytes = block.header.single_vox_offset + bytes_per_voxel_block*(block_shape[0]*block_shape[1]*block_shape[2])
+            unread_split = None
+
+            for split_name in legend:
+               
+                print "Remaining mem: ", remaining_mem 
+                split_name = split_name.strip()
+
+                if unread_split is not None:
+                    unread_split, remaining_mem, read_time = insert_in_dict(data_dict, unread_split, remaining_mem, bytes_per_voxel, y_size, z_size)
+
+                if unread_split is None:    
+                    unread_split, remaining_mem, read_time = insert_in_dict(data_dict, split_name, remaining_mem, bytes_per_voxel, y_size, z_size)
                 
-                if data is None:
-                    data = block.get_data()
-                else:
-                    try:
-
-                        #concatenate data on the right axis
-                        if prev_block_start[1] == block_pos[-2] and prev_block_start[2] == block_pos[-1]:
-                            data = np.concatenate((data,block.get_data()))
-                        elif prev_block_start[1] != block_pos[-2] and prev_block_start[2] == block_pos[-1]:
-                            data = np.concatenate((data,block.get_data()), 1)
-                        elif prev_block_start[2] != block_pos[-1] and prev_block_start[1] == block_pos[-2]:
-                            data = np.concatenate((data,block.get_data()), 2)
-                        else:
-                            raise Exception('Unable to concatenate data')
-                    except Exception as e:
-                        print 'ERROR: ', e
-                        sys.exit(1)
-                        
-                remaining_mem = remaining_mem - (block_bytes)
+                total_read += read_time
 
                 #cannot load another block into memory, must write slice
-                if remaining_mem / block_bytes < 1:
+                if remaining_mem <= 0:
```

Comment:
more than mem is technically not read. it's just i look at the size of next block and deduct it's size from remaining_mem. If  remaining mem is then found to be less than zero, the block will not be read, but instead stored in unread_split to be read during next cluster read. 

I can, however, put an if statement in insert_in_dict along the lines of `if split_bytes > remaining_mem ` , but then I'll have to make some modifications to cluster_reads such that it writes even though remaining_mem > 0  (perhaps something like `if remaining_mem == 0 or unread_split is not None`). How does that sound?